GVFS/GVFS.Common/GVFSEnlistmentHealthCalculator.cs

```diff
@@ -0,0 +1,258 @@
+
+            Dictionary<string, decimal> mostHydratedDirectories = new Dictionary<string, decimal>();
+
+            // Map directory names to the corresponding health data from gitTrackedFilesDirectoryTally and hydratedFilesDirectoryTally
+            foreach (KeyValuePair<string, int> pair in gitTrackedItemsDirectoryTally)
+            {
+                if (hydratedFilesDirectoryTally.TryGetValue(pair.Key, out int hydratedFiles))
+                {
+                    // In-lining this for now until a better "health" calculation is created
+                    // Another possibility is the ability to pass a function to use for health (might not be applicable)
+                    mostHydratedDirectories.Add(pair.Key, this.CalculateHealthMetric(hydratedFiles, pair.Value));
+                }
+                else
+                {
+                    mostHydratedDirectories.Add(pair.Key, 0);
+                }
+            }
+
+            // Generate and return a data object
+            return new GVFSEnlistmentHealthData(
+                parentDirectory,
+                gitTrackedItemsCount,
+                placeholderCount,
+                modifiedPathsCount,
+                this.CalculateHealthMetric(placeholderCount + modifiedPathsCount, gitTrackedItemsCount),
+                mostHydratedDirectories.OrderByDescending(kp => kp.Value).ToList());
+        }
+
+        /// <summary>
+        /// Take a string representing a file path on system and pull out the upper-most directory from it as a string, or GVFSConstants.GitPathSeparator if it is in the root
+        /// </summary>
+        /// <param name="path">The path to a file to parse for the top level directory containing it</param>
+        /// <returns>A string containing the top level directory from the provided path, or GVFSConstants.GitPathSeparator if the path is for an item in the root</returns>
+        private string ParseTopDirectory(string path)
+        {
+            int whackLocation = path.IndexOf(GVFSConstants.GitPathSeparator);
+            if (whackLocation == -1)
+            {
+                return GVFSConstants.GitPathSeparatorString;
+            }
+
+            return path.Substring(0, whackLocation);
+        }
+
+        /// <summary>
+        /// Categorizes a list of paths given as strings by mapping them to the top level directory in their path
+        /// Modifies the directoryTracking dictionary to have an accurate count of the files underneath a top level directory
+        /// </summary>
+        /// <remarks>
+        /// The distinction between files and directories is important --
+        /// If the path to a file doesn't contain a GVFSConstants.GitPathSeparator, then that means it falls within the root
+        /// However if a directory's path doesn't contain a GVFSConstants.GitPathSeparator, it doesn't count towards its own hydration

```

Comment:
> However if a directory's path doesn't contain a GVFSConstants.GitPathSeparator, it doesn't count towards its own hydration

I'm having trouble understanding this comment with all of its negatives.  What happens if a directory's path does contain a GitPathSeparator?  Does it then count toward its own hydration?  (if so, consider updating this comment to "does" statements rather than "doesn't" statements).

Reply:
Yeah that's a pretty unclear comment. I'll clarify here and in the code: the only path that doesn't contain a `GVFSConstants.GitPathSeparator` at this point is one which is a top level directory. Imagine a repository: `src`, `src/foo`, `src/foo/f1.txt`, `src/foo/f2.txt`, `src/foo/f3.txt`. If the only thing in the placeholder list is `foo` (which has been normalized to contain no `/`'s in it), `foo`'s hydration should still be `0` since none of the things it contains are hydrated.

Reply:
tldr; folders that are hydrated don't contribute to their own hydration score (EG they can have 0% hydration - The flip side is that hydrated folders which appear on the placeholder list can have a hydration score of 100%)